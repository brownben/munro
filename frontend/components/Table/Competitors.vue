<script lang="ts" setup>
import type { PropType } from 'vue'
import type { Filters } from '~/utils/filter'
import type { CompetitorWithAgeGender } from '~~/utils/ageClass'

const route = useRoute()
const router = useRouter()

const props = defineProps({
  competitors: {
    type: Array as PropType<CompetitorWithAgeGender[]>,
    required: true,
  },
  filters: { type: Object as PropType<Filters>, required: true },
  eligibility: {
    type: Boolean,
    required: false,
    default: false,
  },
})

type Column = keyof CompetitorWithAgeGender
const ascending = ref(route.query.ascending === 'true')
const activeColumn = ref<Column>(
  (queryToString(route.query.sort_by) as Column) || 'id',
)

const sortedResults = computed(() =>
  props.competitors
    .filter(matchingResults(props.filters))
    .sort(byProperty(ascending.value, activeColumn.value)),
)

const changeSortPreference = (property: Column) => {
  if (property !== activeColumn.value) ascending.value = false
  else ascending.value = !ascending.value
  activeColumn.value = property
}

const ariaSorted = computed(
  (): Partial<Record<Column, 'ascending' | 'descending'>> => ({
    [activeColumn.value]: ascending.value ? 'descending' : 'ascending',
  }),
)

watch([activeColumn, ascending], () => {
  router.replace({
    query: {
      ...route.query,
      sort_by: activeColumn.value,
      ascending: ascending.value ? 'true' : 'false',
    },
  })
})
</script>
<template>
  <table v-if="sortedResults.length > 0" class="w-full">
    <thead>
      <tr
        class="h-14 border-b-2 border-gray-200 text-left text-gray-600 md:h-12 dark:border-gray-400 dark:text-gray-300"
      >
        <th class="text-center" :aria-sort="ariaSorted?.id">
          <button
            type="button"
            class="rounded px-1 font-medium ring-main-200 focus:outline-none focus-visible:ring"
            @click="changeSortPreference('id')"
          >
            Id
            <TableArrows
              :active="activeColumn == 'id'"
              :ascending="ascending"
            />
          </button>
        </th>
        <th :aria-sort="ariaSorted?.name">
          <button
            type="button"
            class="rounded px-2 font-medium ring-main-200 focus:outline-none focus-visible:ring"
            @click="changeSortPreference('name')"
          >
            <span class="sm:hidden">Athlete</span>
            <span class="hidden sm:inline-block">Name</span>
            <TableArrows
              :active="activeColumn == 'name'"
              :ascending="ascending"
            />
          </button>
        </th>
        <th
          class="hidden text-center sm:table-cell"
          :aria-sort="ariaSorted?.age_class"
        >
          <button
            type="button"
            class="rounded px-1 font-medium ring-main-200 focus:outline-none focus-visible:ring"
            @click="changeSortPreference('age_class')"
          >
            Class
            <TableArrows
              :active="activeColumn == 'age_class'"
              :ascending="ascending"
            />
          </button>
        </th>
        <th
          class="hidden text-center md:table-cell"
          :aria-sort="ariaSorted?.club"
        >
          <button
            type="button"
            class="rounded px-1 font-medium ring-main-200 focus:outline-none focus-visible:ring"
            @click="changeSortPreference('club')"
          >
            Club
            <TableArrows
              :active="activeColumn == 'club'"
              :ascending="ascending"
            />
          </button>
        </th>
        <th
          v-if="eligibility"
          class="hidden px-1 text-center font-medium sm:table-cell"
          :aria-sort="ariaSorted?.club"
        >
          Eligible?
        </th>
        <th></th>
      </tr>
    </thead>
    <tbody>
      <tr
        v-for="competitor of sortedResults"
        :key="competitor.id"
        class="border-t border-gray-200 text-gray-800 transition hover:bg-main-50 dark:border-gray-700 dark:text-gray-200 dark:hover:bg-gray-800"
      >
        <td class="py-3 text-center">
          {{ competitor.id }}
        </td>
        <td class="px-2 py-2">
          <NuxtLink
            :to="`/competitors/${competitor.id}`"
            class="inline-block ring-main-200 hover:text-main-800 focus-visible:ring"
          >
            <span
              class="block font-medium leading-tight text-gray-900 underline decoration-gray-300 decoration-1 underline-offset-2 dark:text-gray-100 dark:decoration-gray-500"
            >
              {{ competitor.name }}
            </span>
            <span
              class="text-sm leading-tight text-gray-500 sm:hidden dark:text-gray-400"
            >
              <template v-if="competitor.age_class">
                {{ competitor.age_class }} &nbsp;
              </template>
              {{ competitor.club }}
            </span>
          </NuxtLink>
        </td>
        <td class="hidden text-center sm:table-cell">
          {{ competitor.age_class || ' ' }}
        </td>
        <td class="hidden text-center md:table-cell">
          {{ competitor.club || ' ' }}
        </td>
        <td v-if="eligibility" class="hidden text-center sm:table-cell">
          {{ competitor.eligible }}
        </td>
        <td>
          <div class="flex flex-wrap justify-around gap-2">
            <slot :competitor="competitor"></slot>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
  <div v-else class="px-6 lg:px-0">
    <p class="text-xl font-extrabold text-gray-600 dark:text-gray-300">
      No competitors found
    </p>
    <p class="text-gray-500 dark:text-gray-400">
      Try changing the filters selected.
    </p>
  </div>
</template>
