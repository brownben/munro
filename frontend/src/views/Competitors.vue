<!--
  Competitors

  Displays list of competitors
-->

<template>
  <div class="view">
    <vue-headful
      title="Munro - Competitors"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <h1 class="text-main text-3xl font-normal font-heading mb-2">Competitors</h1>
    <div class="card">
      <h2 class="text-2xl font-heading">Actions</h2>

      <div>
        <router-link to="/create-competitor" class="button">Add Competitor</router-link>
        <router-link to="/competitors/merge" class="button">Merge Competitors</router-link>
        <router-link to="/results/transfer" class="button">Transfer Result</router-link>
      </div>

      <div class="text-left w-full">
        <dropdown-input
          v-model="league"
          :list="leagues.map(league => league.name)"
          label="League:"
        />
      </div>
    </div>

    <table v-if="filteredCompetitors.length > 0" class="table my-4 table-fixed">
      <thead>
        <tr>
          <th @click="sortBy('name')">
            <p>Name</p>
            <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'name'" />
          </th>
          <th @click="sortBy('club')">
            <p>Club</p>
            <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'club'" />
          </th>
          <th @click="sortBy('ageClass')">
            <p>Class</p>
            <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'ageClass'" />
          </th>
          <th @click="sortBy('course')">
            <p>Course</p>
            <up-down-arrow :ascending="ascendingSort" :active="sortedBy === 'course'" />
          </th>
        </tr>
      </thead>
      <tbody is="transition-group" name="fade">
        <tr
          v-for="competitor of sortedCompetitors"
          :key="competitor.name+competitor.league+competitor.course"
          :class="{ striped: sortedCompetitors.indexOf(competitor) % 2 === 0 }"
          @click="$router.push('/competitors/' + competitor.id + '/edit')"
        >
          <td>{{ competitor.name }}</td>
          <td>{{ competitor.club }}</td>
          <td>{{ competitor.ageClass }}</td>
          <td>{{ competitor.course }}</td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<script>
import axios from 'axios'
import UpDownArrow from '@/components/UpDownArrows'
import DropdownInput from '../components/DropdownInput.vue'

export default {
  components: {
    UpDownArrow: UpDownArrow,
    DropdownInput: DropdownInput,
  },

  data: () => ({
    league: '',
    leagues: [],
    competitors: [],
    ascendingSort: false,
    sortedBy: 'name',
  }),

  computed: {
    filteredCompetitors: function () {
      return this.competitors.filter(competitor => competitor.league === this.league)
    },

    sortedCompetitors: function () {
      return this.sort(
        this.filteredCompetitors,
        this.sortedBy,
        this.ascendingSort)
    },
  },

  created: async function () {
    if (this.$route.params.league && this.$route.params.league !== '') this.league = this.$route.params.league
    this.getLeagues()
    this.getCompetitors()
  },

  methods: {
    getCompetitors: function () {
      return axios.get('/api/competitors')
        .then(response => { this.competitors = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching Competitor Details'))
    },

    sort: function (array, property, ascending = true) {
      // Selection Sort using Single List for Sorting Results

      const sortFunction = (a, b) => {
        if (a[property] === b[property]) return 0
        else if (a[property] === null) return 1
        else if (b[property] === null) return -1
        else if (a[property] < b[property]) return 1
        else return -1
      }

      if (ascending) return array.sort(sortFunction)
      else return array.sort(sortFunction).reverse()
    },

    sortBy: function (sortBy) {
      // Change what property it is sorted by
      // If it is a different property, make it sort ascending else change direction of sort
      if (sortBy !== this.sortedBy) this.ascendingSort = false
      else this.ascendingSort = !this.ascendingSort
      this.sortedBy = sortBy
    },

    getLeagues: function () {
      return axios.get('/api/leagues')
        .then(response => { this.leagues = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching List of Leagues'))
    },
  },
}
</script>

