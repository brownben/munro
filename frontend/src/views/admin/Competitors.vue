<!--
  Competitors

  Displays list of competitors
-->

<template>
  <Layout has-mobile-sub-title>
    <Meta
      :head="{
        meta: { name: 'robots', content: 'noindex' },
      }"
      title="Munro - Competitors"
      description
    />

    <template #title>
      <h1 class="text-3xl font-bold leading-tight font-heading">
        <router-link
          :to="'/leagues/' + $route.params.league"
          class="text-xl text-main-700 md:text-3xl"
        >
          {{ $route.params.league }}
        </router-link>
        <span class="hidden ml-2 mr-3 md:inline-block">-</span>
        <span class="block text-4xl md:text-3xl md:inline-block">
          Competitors
        </span>
      </h1>
    </template>

    <div class="col-span-2 card card-color-dark">
      <h2 class="text-3xl font-bold text-white font-heading">Admin Actions</h2>

      <div class="w-full">
        <router-link to="/competitors/create" class="button button-white">
          Add Competitor
        </router-link>
        <router-link to="/competitors/merge" class="button button-white">
          Merge Competitors
        </router-link>
        <router-link to="/results/transfer" class="button button-white">
          Transfer Result
        </router-link>
        <router-link to="/results/manual" class="button button-white">
          Manual Points
        </router-link>
      </div>
    </div>

    <table
      v-if="competitors.length > 0"
      class="table w-full col-span-2 my-4 border-collapse table-fixed"
    >
      <thead>
        <tr
          class="transition duration-300 bg-white border-b border-collapse border-main-300 hover:bg-main-200"
        >
          <th @click="changeSortPreference(SortablePropeties.id)">
            <p>Id</p>
            <up-down-arrow
              :ascending="sortPreferences.ascending"
              :active="sortPreferences.by === SortablePropeties.id"
            />
          </th>
          <th
            class="text-left"
            @click="changeSortPreference(SortablePropeties.name)"
          >
            <p>Name</p>
            <up-down-arrow
              :ascending="sortPreferences.ascending"
              :active="sortPreferences.by === SortablePropeties.name"
            />
          </th>
          <th
            class="club"
            @click="changeSortPreference(SortablePropeties.club)"
          >
            <p>Club</p>
            <up-down-arrow
              :ascending="sortPreferences.ascending"
              :active="sortPreferences.by === SortablePropeties.club"
            />
          </th>
          <th
            class="ageClass"
            @click="changeSortPreference(SortablePropeties.ageClass)"
          >
            <p>Class</p>
            <up-down-arrow
              :ascending="sortPreferences.ascending"
              :active="sortPreferences.by === SortablePropeties.ageClass"
            />
          </th>
          <th @click="changeSortPreference(SortablePropeties.course)">
            <p>Course</p>
            <up-down-arrow
              :ascending="sortPreferences.ascending"
              :active="sortPreferences.by === SortablePropeties.course"
            />
          </th>
        </tr>
      </thead>
      <transition-group name="list" tag="tbody">
        <tr
          v-for="competitor of competitors"
          :key="competitor.id"
          class="transition duration-300 bg-white border-collapse hover:bg-main-200"
          :class="{
            'bg-main-50': competitors.indexOf(competitor) % 2 === 0,
          }"
          @click="$router.push(`/competitors/${competitor.id}`)"
        >
          <td class="text-center">
            {{ competitor.id }}
          </td>
          <td>
            <span class="block font-normal text-left md:font-light">
              {{ competitor.name }}
            </span>
            <span class="block text-xs md:hidden">
              <span v-if="competitor.ageClass" class="mr-4">
                {{ competitor.ageClass }}
              </span>
              <span>{{ competitor.club }}</span>
            </span>
          </td>
          <td class="hidden text-center md:table-cell">
            {{ competitor.club }}
          </td>
          <td class="hidden text-center md:table-cell">
            {{ competitor.ageClass }}
          </td>
          <td class="text-center">
            {{ competitor.course }}
          </td>
        </tr>
      </transition-group>
    </table>
  </Layout>
</template>
<script lang="ts">
import Layout from '/@/components/Layout.vue'
import UpDownArrow from '/@/components/UpDownArrows.vue'

export default {
  components: {
    Layout,
    UpDownArrow,
  },
}
</script>
<script lang="ts" setup>
import { ref, watch, computed } from 'vue'

import { toSingleString } from '/@/scripts/typeHelpers'
import { sortCompetitors } from '/@/scripts/sort'

import $router from '/@/router/index'
const { currentRoute: $route } = $router

import { Competitor, getLeagueCompetitors } from '/@/api/competitors'

import {
  SortPreferencesCompetitor as SortPreferences,
  SortablePropetiesCompetitor as SortablePropeties,
} from '/@/scripts/FilterSort.d'

/* Get Data */
const loading = ref(true)
const rawCompetitors = ref<Competitor[]>([])

const getData = async () => {
  const routeParamsLeague = toSingleString($route.value.params.league)

  loading.value = true
  rawCompetitors.value = await getLeagueCompetitors(routeParamsLeague)
  loading.value = false
}

const competitors = computed(() =>
  rawCompetitors.value.sort(sortCompetitors(sortPreferences.value))
)

watch($route, getData, { immediate: true })

export { loading, competitors }

/* Sorting */
const sortPreferences = ref<SortPreferences>({
  ascending: false,
  by: SortablePropeties.position,
})
const changeSortPreference = (property: SortablePropeties) => {
  if (property !== sortPreferences.value.by)
    sortPreferences.value.ascending = false
  else sortPreferences.value.ascending = !sortPreferences.value.ascending
  sortPreferences.value.by = property
}

export { sortPreferences, SortablePropeties, changeSortPreference }
</script>

<style lang="postcss" scoped>
table {
  & td {
    @apply py-2 px-1 font-sans font-light;
  }

  & th {
    white-space: nowrap;
    @apply font-heading select-none font-normal py-2;

    & p {
      @apply inline-block;
    }

    & div {
      @apply inline-block ml-1;
    }
  }
}
</style>
