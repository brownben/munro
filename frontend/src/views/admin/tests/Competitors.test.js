import axios from 'axios'
import { shallowMount } from '@vue/test-utils'
import Competitors from '@/views/admin/Competitors'

jest.mock('axios')
const flushPromises = () =>
  new Promise((resolve) => global.setImmediate(resolve))

beforeEach(() => {
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: [] })
})

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { params: { league: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { params: { league: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Get Competitors - Success', async () => {
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { params: { league: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  flushPromises()
  axios.get.mockResolvedValue({
    data: [
      { name: 1, league: 'a' },
      { name: 2, league: 'a' },
      { name: 3, league: 'a' },
      { name: 4, league: 'b' },
    ],
  })
  await wrapper.vm.getCompetitors()
  expect(wrapper.vm.competitors).toEqual([
    { name: 1, league: 'a' },
    { name: 2, league: 'a' },
    { name: 3, league: 'a' },
    { name: 4, league: 'b' },
  ])
  wrapper.setData({ league: 'a' })
  expect(wrapper.vm.filteredCompetitors).toEqual([
    { name: 1, league: 'a' },
    { name: 2, league: 'a' },
    { name: 3, league: 'a' },
  ])
})

test('Renders Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { params: { league: '' } },
      $messages: { addMessage: mockAddMessage },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  flushPromises()
  axios.get.mockRejectedValue()
  await wrapper.vm.getCompetitors()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith(
    'Problem Fetching Competitor Details'
  )
})

test('Change Sort Order', () => {
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { params: { league: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ sortedBy: 'Age' })
  wrapper.vm.sortBy('Name')
  expect(wrapper.vm.ascendingSort).toBeFalsy()
  expect(wrapper.vm.sortedBy).toEqual('Name')
  wrapper.vm.sortBy('Name')
  expect(wrapper.vm.ascendingSort).toBeTruthy()
  expect(wrapper.vm.sortedBy).toEqual('Name')
})

test('Sort', () => {
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { params: { league: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.sort([{ a: 3 }, { a: 1 }, { a: 2 }], 'a', false)).toEqual([
    { a: 1 },
    { a: 2 },
    { a: 3 },
  ])
  expect(
    wrapper.vm.sort([{ a: 1 }, { a: 3 }, { a: 1 }, { a: 2 }], 'a', true)
  ).toEqual([{ a: 3 }, { a: 2 }, { a: 1 }, { a: 1 }])
  expect(
    wrapper.vm.sort([{ a: 1 }, { a: 3 }, { a: 1 }, { a: 2 }], 'a')
  ).toEqual([{ a: 3 }, { a: 2 }, { a: 1 }, { a: 1 }])
})

test('Get Leagues - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '', id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: [{ league: '1' }, { league: '2' }] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([{ league: '1' }, { league: '2' }])

  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])

  axios.get.mockResolvedValue({
    data: [
      { league: 'a', property: 3 },
      { league: '2', another: 4 },
    ],
  })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([
    { league: 'a', property: 3 },
    { league: '2', another: 4 },
  ])
})

test('Get Leagues - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(Competitors, {
    mocks: {
      $route: { path: '/events/1/edit', params: { league: '', id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Fetching List of Leagues'
  )
})
