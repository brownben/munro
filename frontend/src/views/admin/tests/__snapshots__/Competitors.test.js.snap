// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<layout-stub
  title=""
>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Competitors"
  />
   
  <h1
    class="col-span-2 font-bold text-left text-25xl md:text-3xl font-heading text-main-900"
  >
    <router-link-stub
      class="link text-main-700"
      to="/leagues/"
    >
      
      
    
    </router-link-stub>
     
    <span
      class="hidden ml-2 mr-3 md:inline-block"
    >
      -
    </span>
     
    <span
      class="block text-35xl md:inline-block"
    >
       Competitors
    </span>
  </h1>
   
  <div
    class="col-span-2 card-color-dark"
  >
    <h2
      class="text-3xl font-bold text-white font-heading"
    >
      Admin Actions
    </h2>
     
    <div
      class="w-full"
    >
      <router-link-stub
        class="button-white"
        to="/competitors/create"
      >
        Add Competitor
      </router-link-stub>
       
      <router-link-stub
        class="button-white"
        to="/competitors/merge"
      >
        Merge Competitors
      </router-link-stub>
       
      <router-link-stub
        class="button-white"
        to="/results/transfer"
      >
        Transfer Result
      </router-link-stub>
       
      <router-link-stub
        class="button-white"
        to="/results/manual"
      >
        Manual Points
      </router-link-stub>
    </div>
  </div>
   
  <!---->
</layout-stub>
`;
